pageserver/src/tenant/timeline.rs

```diff
@@ -3604,41 +3602,52 @@ impl Timeline {
                         return;
                     };
-                    num_frozen_layers = lm.frozen_layers.len();
-                    frozen_layer_total_size = lm
+                    let l0_count = lm.level0_deltas().len();
+                    let frozen_count = lm.frozen_layers.len();
+                    let frozen_size: u64 = lm
                         .frozen_layers
                         .iter()
                         .map(|l| l.estimated_in_mem_size())
-                        .sum::<u64>();
-                    lm.frozen_layers.front().cloned()
-                    // drop 'layers' lock to allow concurrent reads and writes
+                        .sum();
+                    let layer = lm.frozen_layers.front().cloned();
+                    (layer, l0_count, frozen_count, frozen_size)
+                    // drop 'layers' lock
                 };
-                let Some(layer_to_flush) = layer_to_flush else {
+                let Some(layer) = layer else {
                     break Ok(());
                 };
-                if num_frozen_layers
-                    > std::cmp::max(
-                        self.get_compaction_threshold(),
-                        DEFAULT_COMPACTION_THRESHOLD,
-                    )
-                    && frozen_layer_total_size >= /* 128 MB */ 128000000
-                {
-                    tracing::warn!(
-                        "too many frozen layers: {num_frozen_layers} layers with estimated in-mem size of {frozen_layer_total_size} bytes",
-                    );
-                }
-                match self.flush_frozen_layer(layer_to_flush, ctx).await {
-                    Ok(this_layer_to_lsn) => {
-                        flushed_to_lsn = std::cmp::max(flushed_to_lsn, this_layer_to_lsn);
+
+                // Backpressure if compaction can't keep up. At 2x the compaction threshold, we
+                // delay flushes by the flush duration (i.e. each flush takes 2x as long). At 3x the
+                // compaction threshold, we stall flushes.
+                let compaction_threshold = max(
+                    self.get_compaction_threshold(),
+                    DEFAULT_COMPACTION_THRESHOLD,
+                );
+                let delay_threshold = 2 * compaction_threshold;
+                let stall_threshold = 3 * compaction_threshold;
+
+                if l0_count >= stall_threshold {
+                    // TODO: don't log for every iteration.
+                    warn!("stalling flush for compaction: {l0_count} L0 layers, {frozen_count} frozen layers of {frozen_size} bytes");
+                    // TODO: can we get compaction to notify us when it's done?
+                    tokio::select! {
+                        _ = tokio::time::sleep(Duration::from_millis(100)) => {},
+                        _ = self.cancel.cancelled() => continue,
                     }
+                    continue;
```

Comment:
I wonder if it is possible that stalling flushing will cause the in-memory layer to grow very large. Probably we should stall at an earlier place of the ingestion path? (i.e., after we get the data from walreceiver and before we write them into the in-memory layer)

Reply:
Yep. I added 94a8a0783a28bc9275e4779ee59dd8e36265aee7 to propagate the flush backpressure up into ephemeral layer rolls. That should stall WAL ingestion.